### PR TITLE
Feature/issue 1142 Remove padding from OpenCL general matrix multiply

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

This PR removes the use of padding for the input and output matrices in the general matrix multiplication and multiply transpose. Padding was used mainly to simplify the code and provide some performances boosts with large matrices. 

Benchmarks shown below indicate that removing the padding  reduces the execution time for smaller matrices to about 45% and that the execution time increases for 5% at maximum for large matrices. But for that 5% of execution time with padding we actually more than double the memory footprint. Which to me is not a worthy sacrifice. 

The performance loss is smaller for rectangular input matrices that suppose to come up more often in matrix multiplications in Stan (as far as I know, there was some discussion on Discourse some time ago). 

At square matrix multiplication with N=10000 the loss is 1.5%, at rectangular matrix multiply [4000, 24000] x [24000, 4000] that has similar sized matrices, the loss is around 0.8% (see [rect_mult_padding_perf.txt](https://github.com/stan-dev/math/files/3000540/rect_mult_padding_perf.txt)). 

Removing padding will also enable easier optimizations for multiplications of rectangular matrices where the common dimension is larger: [N, K] x [K, M] where K >>> N,M; the extreme cases of this are matrix x row_vector or vector x matrix, otherwise for example [500, 10000] x [10000 x 10]. I managed to get around 20%-50% reduction in those cases (will open once this PR gets merged, if it does).

Will post the same results on a Windows machine with and AMD GPU. The V100 AWS gives comparable results to the presented that were obtained on an Ubuntu system with the TitanXP.

Benchmark results:
```
Padding:
------------------------------------------------------------
Benchmark                     Time           CPU Iterations
------------------------------------------------------------
BM_square_mult/100       117878 ns     117880 ns       5403
BM_square_mult/200       172075 ns     172073 ns       4028
BM_square_mult/300       687443 ns     686188 ns       1003
BM_square_mult/400      1287351 ns    1285840 ns        540
BM_square_mult/500      1704510 ns    1701650 ns        409
BM_square_mult/600      2479740 ns    2477277 ns        280
BM_square_mult/700      3135085 ns    3131587 ns        224
BM_square_mult/800      4299533 ns    4297548 ns        163
BM_square_mult/900      6151174 ns    6145165 ns        110
BM_square_mult/1000     7500103 ns    7498104 ns         92
BM_square_mult/2000    49935540 ns   49907178 ns         14
BM_square_mult/3000   158138707 ns  158139698 ns          5
BM_square_mult/4000   377430860 ns  377441717 ns          2
BM_square_mult/5000   703299608 ns  703224491 ns          1
BM_square_mult/6000  1205413056 ns 1205430843 ns          1
BM_square_mult/7000  1899133406 ns 1898768774 ns          1
BM_square_mult/8000  2821377736 ns 2821007537 ns          1
BM_square_mult/9000  4041553169 ns 4040393951 ns          1
BM_square_mult/10000 5583241541 ns 5581643557 ns          1
BM_square_mult/12000 9695023441 ns 9693572307 ns          1
BM_square_mult/14000 15454004596 ns 15451043311 ns          1

No padding:
------------------------------------------------------------
Benchmark                     Time           CPU Iterations
------------------------------------------------------------
BM_square_mult/100        50532 ns      50498 ns      11740
BM_square_mult/200       101143 ns     101145 ns       6830
BM_square_mult/300       245366 ns     245226 ns       2849
BM_square_mult/400       624087 ns     623715 ns       1095
BM_square_mult/500      1055132 ns    1054575 ns        657
BM_square_mult/600      1544709 ns    1543279 ns        450
BM_square_mult/700      2306728 ns    2305893 ns        304
BM_square_mult/800      3314078 ns    3311070 ns        212
BM_square_mult/900      5022115 ns    5021233 ns        136
BM_square_mult/1000     6382371 ns    6377514 ns        108
BM_square_mult/2000    47042385 ns   47042352 ns         15
BM_square_mult/3000   156150004 ns  155448507 ns          5
BM_square_mult/4000   360692166 ns  360493414 ns          2
BM_square_mult/5000   711667850 ns  711671852 ns          1
BM_square_mult/6000  1217096868 ns 1216713130 ns          1
BM_square_mult/7000  1924372568 ns 1923841980 ns          1
BM_square_mult/8000  2963520069 ns 2963505009 ns          1
BM_square_mult/9000  4354807345 ns 4354897041 ns          1
BM_square_mult/10000 5857150628 ns 5857236273 ns          1
BM_square_mult/12000 9789256097 ns 9789489058 ns          1
BM_square_mult/14000 15648197309 ns 15647315947 ns          1
```

## Tests

Run `/opencl/multiply_test.cpp` and `/opencl/multiply_transpose.cpp`. 

This PR expands test coverage for triangular multiplication when the input is not zeroed-out.

## Side Effects

Some performance loss in larger matrix multiplications for the sake of smaller memory footprint.
The kernel for general matrix multiply is also a bit more complex now.

## Checklist

- [x] Math issue #1142 

- [x] Copyright holder: Rok Češnovar, Univ. of Ljubljana

- [x] the basic tests are passing

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
Forgot to include the benchmark file: https://github.com/rok-cesnovar/perf-math/blob/master/matrix_test_padding.cpp

The timing was for the mutliplication only, without data transfer, so the overall impact is even smaller.

**Comment by rok-cesnovar** [comment]
I wasnt able to setup Google Benchmark on the Windows machine, so here are the C++ timings for the Windows+AMD R9 Fury GPU:


N | padding | nopadding | performance loss
-- | -- | -- | --
100 | 339.08 | 149.45 | -0.5592485549
200 | 438.88 | 279.36 | -0.3634706526
300 | 578.4 | 439.08 | -0.2408713693
400 | 967.8 | 738.02 | -0.2374250878
500 | 1627.92 | 1348.95 | -0.1713659148
600 | 2208.79 | 2167.98 | -0.01847617927
700 | 3369.35 | 3206.99 | -0.04818733584
800 | 4600.92 | 4502.23 | -0.02145005781
900 | 6471.14 | 6549.32 | 0.01208133343
1000 | 8396.29 | 8401.22 | 0.0005871640927
2000 | 59716.9 | 61234.1 | 0.02540654321
3000 | 197550 | 205648 | 0.04099215389
4000 | 460437 | 480222 | 0.04297004802
5000 | 917789 | 979953 | 0.06773234371

File: stan/math/opencl/kernels/matrix_multiply.hpp [hunk 1]
```diff
@@ -79,12 +79,33 @@ static const char* matrix_multiply_kernel_code = STRINGIFY(
           // For the tiles on the diagonal we can ignore the values over
           // the diagonal if the matrix is lower triangular or under
           // the diagonal if the matrix is upper triangular
-          A_local[thread_block_col + w * THREAD_BLOCK_SIZE_COL]
-                 [thread_block_row]
-              = A[(tiled_j + w * THREAD_BLOCK_SIZE_COL) * M + i];
-          B_local[thread_block_col + w * THREAD_BLOCK_SIZE_COL]
-                 [thread_block_row]
-              = B[(j + w * THREAD_BLOCK_SIZE_COL) * K + tiled_i];
+          const A_temp_j = tiled_j + w * THREAD_BLOCK_SIZE_COL;
```
Comment (seantalts) [inline stan/math/opencl/kernels/matrix_multiply.hpp]: I'm not sure what the temp things are here, or why they're temp - are there any other words you could use to describe them or put in a comment nearby? is it basically just the current column index? Would a name like `B_curr_col` or `B_curr_j` work? I'm not really sure what is typical in code like this. I was originally going to recommend splitting this out into a function but abstraction is so annoying in OpenCL, the function would have like 10 arguments.

File: stan/math/opencl/kernels/matrix_multiply.hpp [hunk 2]
```diff
@@ -99,7 +120,10 @@ static const char* matrix_multiply_kernel_code = STRINGIFY(
       // save the values
       for (int w = 0; w < WORK_PER_THREAD; w++) {
         // each thread saves WORK_PER_THREAD values
-        C[(j + w * THREAD_BLOCK_SIZE_COL) * M + i] = acc[w];
+        // padded threads should not access C
```
Comment (seantalts) [inline stan/math/opencl/kernels/matrix_multiply.hpp]: is it padded threads? or that threads shouldn't access C if they'd be accessing something unpadded? If it was the former I might expect this check to not use `w` and take place outside of the immediate for loop. But probably you will have threads that are part padded, part not right?

File: stan/math/opencl/kernels/matrix_multiply.hpp (hunk 1 above)
Comment (rok-cesnovar) [reply stan/math/opencl/kernels/matrix_multiply.hpp]: Yep, B_curr_col would work and definitely be better. Will rename these.

File: stan/math/opencl/kernels/matrix_multiply.hpp (hunk 2 above)
Comment (rok-cesnovar) [reply stan/math/opencl/kernels/matrix_multiply.hpp]: Yeah this is due to threads that part padded part not.

File: test/unit/math/opencl/multiply_test.cpp [hunk 3]
```diff
@@ -265,9 +264,8 @@ TEST(MathMatrix, upper_tri_rect_multiply_big_rect) {
   auto m2 = stan::math::matrix_d::Random(size, size * 3).eval();
   stan::math::matrix_d m3_cl_res(size, size * 3);
 
-  m1.triangularView<Eigen::StrictlyLower>().setZero();
-
   stan::math::matrix_cl m11(m1);
+  m1.triangularView<Eigen::StrictlyLower>().setZero();
```
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply test/unit/math/opencl/multiply_test.cpp]: For better test coverage. We transfer non zeroed out matrix in order to test that the kernels really dont access elements above/under the diagonal. So we dont rely that the input will have zeros on the not used part of the matrix.

[the remark this answers is not in the document]
Comment (seantalts) [reply test/unit/math/opencl/multiply_test.cpp]: gotcha! Makes sense. Didn't see it was initialized to random above. 👍

File: stan/math/opencl/kernels/matrix_multiply.hpp (hunk 2 above)
Comment (seantalts) [reply stan/math/opencl/kernels/matrix_multiply.hpp]: Would you mind also modifying the comment to say something like that? Thanks!

Comment (rok-cesnovar) [reply stan/math/opencl/kernels/matrix_multiply.hpp]: Added, sorry for jumping the gun on clicking resolve on this one.

**Pull request merged**